enhancements/ingress/global-options-enable-hsts.md

```diff
@@ -0,0 +1,457 @@
+`reencrypt` routes that match the `haproxy.router.openshift.io/hsts_header` annotation:
+```gotemplate
+{{- /* hsts header in response: */}}
+ {{- $hstsOptionalTokenPattern := `(?:includeSubDomains|preload)` }}
+ {{- $hstsPattern := printf `(?:%[1]s[;])*max-age=(?:\d+|"\d+")(?:[;]%[1]s)*`  $hstsOptionalTokenPattern -}}
+...
+{{- if matchValues (print $cfg.TLSTermination) "edge" "reencrypt" }}
+    {{- with $hsts := firstMatch $hstsPattern (index $cfg.Annotations "haproxy.router.openshift.io/hsts_header") }}
+  http-response set-header Strict-Transport-Security {{$hsts}}
+    {{- end }}{{/* hsts header */}}
+  {{- end }}{{/* is "edge" or "reencrypt" */}}
+```
+
+#### Enable HSTS for all routes globally, based on the cluster's ingress domain
+To add the header to all HTTP responses from an `edge` or `reencrypt` route, add a new router environment
+variable called `ROUTER_DEFAULT_HSTS`.
+
+Add a new type `IngressControllerHSTSHeaderPolicy` to the API in
+`openshift/api/blob/master/operator/v1/types_ingress.go`, to capture the configuration of the HSTS Policy.
+```go
+// These are strings instead of enums so that they can be passed in easily from the HAProxy template
+const (
+    AllHSTSHeaderPolicyScope         = "All"
+    LimitedHSTSHeaderPolicyScope     = "Limited"
+)
+
+// HSTSDirectiveType is an optional directive that directs HSTS Policy configuration.
+//+kubebuilder:validation:Enum=preload;includeSubDomains
+type HSTSDirectiveType string
+const(
+    // PreloadHSTSDirective directs the client to include this host in its host preload list so that
+    // it never needs to do an initial load to get the HSTS header (note that this is not defined
+    // in RFC 6797 and is therefore client implementation-dependent).
+    PreloadHSTSDirective             HSTSDirectiveType = "preload"
+
+    // IncludeSubDomainsHSTSDirective means the HSTS Policy applies to any subdomains of the host's domain name
+    IncludeSubDomainsHSTSDirective   HSTSDirectiveType = "includeSubDomains"
+)
+
+type IngressControllerHSTSHeaderPolicy {
+	// maxAgeSeconds is the delta time range in seconds during which the host is regarded as an HSTS host.
+	// If set to 0, it negates the effect and the host is no longer regarded as an HSTS host.
+	// maxAgeSeconds is a time-to-live value, and if it is not refreshed, the HSTS policy for this host
+	// will eventually expire.
+	// +required
+    MaxAgeSeconds       int32 `json:"maxAgeSeconds"`
+
+    // HSTSDirectives is a list of optional HSTS directives that direct HSTS Policy configuration.
+    HSTSDirectives      []HSTSDirectiveType `json:"hstsDirectives,omitempty"`
+
+    // scope indicates whether this header policy is enabled for routes that have hosts with domains
+    // in the list specified in inScopeDomains (limited scope) or for all routes that have hosts under the
+    // cluster ingress domain (all scope).  If global HSTS is desired, set the scope field to "all".  The
+    // default is "all".
```

Comment:
We don't want to start enabling HSTS everywhere when a user upgrades a cluster, so either the default should be some new "None" value, or the default should be "Limited" (which effectively means "None" as `inScopeDomains` defaults to empty).

Reply:
You are right.  The default will be none if Scope is not specified.